### PR TITLE
feat(cli): support heterogeneous prefill and decode hardware

### PR DESCRIPTION
**Pull request by jasonqinzhou**
## Why and what changed

A YAML file could describe heterogeneous P/D hardware through the Sweeper SDK after #186, but unified `aisimulate recommend` and `predict` rejected per-worker hardware. Users also could not save and replay those recommendations through the unified CLI.

Add optional `engine.workers.prefill.hardware` and `engine.workers.decode.hardware`. Each omitted role inherits `engine.hardware`. Recommendation forwards these values to the existing Sweeper implementation and preserves them in selected prediction YAML; prediction uses each effective SKU for timing, KV capacity, and metadata. P/D still shares one model, backend, backend version, and total GPU budget. The existing Router AIC hardware guard now also protects prediction.

Migration section 4.8 now includes a complete two-hardware YAML, `recommend` and `predict` commands, and an expected result. This PR includes the now-merged #200 as its base and preserves its numbered guide structure.

## Review map

- Risk level: medium.
- Start with `config/engine.py` → `recommend.py` → `compiler.py`; `tests/e2e/test_unified_cli_engine.py` exercises the real saved-YAML consumer.
- Public or serialized contract changed: two optional concrete worker hardware fields. Hardware domains, mixed P/D models/backends/versions, AFD companion overrides, and deployment generation are outside this change.
- Compatibility: omitted overrides preserve the existing CLI YAML serialization and compiler defaults. Explicit overrides without a pinned backend version require a common resolved version. Aggregated workers and AFD companions reject overrides.

## Evidence

- 469 focused configuration, CLI, native replay, EPD, and Sweeper tests passed in one combined run (109.85 s) at `0dd17f819bc2dae1ad205641d028c37b4899214f`. The final closeout also merges current `main` at `be8f6974`, rejects whitespace-padded worker hardware, and corrects the remaining contradictory guide paragraph.
- Native H200/GB200 Llama 3.1 8B `recommend → recommendations/0001.yaml → predict`: six completed requests; TTFT, TPOT, and output throughput reproduce candidate metrics. Uses real op-level profiles and inferred KV capacity, not fixed timing.
- Native analytical EPD round trips also cover GB200 decode with H200 prefill/encoder.
- Negative/boundary cases: missing/one/both role overrides; matching overrides distinct from fallback; `auto` fallback; mixed aggregated/P/D search; invalid worker SKUs/domains; AFD/aggregated rejection; inconsistent implicit versions; stale Router AIC hardware.
- Repository Fast CI Ruff scope and formatting pass. Checked 90 local documentation links/anchors, 46 Bash blocks, 32 YAML blocks, and seven complete CLI configurations. Packaged legal files match.
- Final closeout at `3c3a4fa342e3d50717cf95ca99e83a14fd5104cc`: 236 focused configuration, analytical EPD, native CLI round-trip, and Sweeper tests passed in 95.65 s using an isolated environment and a native extension built from this checkout. The expanded hardware matrix covers both roles and both public schemas, including leading/trailing whitespace, padded `auto`, and padded concrete SKUs. Unknown worker SKUs now fail during prediction compilation for explicit and implicit backend versions, while configured custom system paths remain supported. The public CLI returns a configuration error (exit 2) instead of a later KV-cache runtime error. Ruff, formatting, whitespace checks, and guide-YAML validation passed.
- Codex reviewed commit: `3c3a4fa342e3d50717cf95ca99e83a14fd5104cc`, including the complete PR diff against `main` at `be8f69749589aea5a376083436e054d75810d379`. No remaining actionable finding; the hardware validation and documentation conversations are fixed and resolved.
- [Fast CI](https://github.com/ai-dynamo/aisimulate/actions/runs/34909448762), DCO, CODEOWNERS, and CodeRabbit passed on the final head. [Full CI](https://github.com/ai-dynamo/aisimulate/actions/runs/34909763821) passed with `expected_sha=3c3a4fa342e3d50717cf95ca99e83a14fd5104cc`: 56 successful jobs and the expected release-only wheel-staging skip. GitHub reports `APPROVED`, all four conversations are resolved, and the branch includes current `main` at `be8f6974`.

## Modeling or data provenance

No performance formulas, profiles, or expected-value goldens change. The native replay comparison checks persistence of the evaluated configuration; it is not a silicon accuracy claim.

## Tracking

- Related: #186, #200; AIC-1772.

### CODE REVIEW

File: docs/cli/user-guide.md [hunk 1]
```diff
@@ -922,6 +923,15 @@ engine:
 identifier. Language-worker roles in aggregated or disaggregated mode use the same hardware;
```
Comment (jasonqinzhou) [inline docs/cli/user-guide.md]: P2: The preceding paragraph still says language-worker roles always use the same hardware. That contradicts the per-role overrides added immediately below and can make users conclude the new YAML remains unsupported. Update this inherited sentence to describe the shared fallback and optional P/D overrides.

Comment (jasonqinzhou) [reply docs/cli/user-guide.md]: Fixed in 91c75b86cc335a355e8cfc1404f8764d12ccc4b9. The guide now describes fallback inheritance and optional P/D overrides. Rechecked the guide links and complete YAML examples.

**Comment by jasonqinzhou** [comment]
Codex self-review completed at `91c75b86cc335a355e8cfc1404f8764d12ccc4b9` against `main` at `9ad041ea989878b2b6adae5eb1ac273bdbb71b27`.

Traced worker hardware through public validation, Sweeper role enumeration and candidate materialization, saved YAML, compiler timing/KV descriptors, native replay, and the Router AIC compatibility guard. One stale guide statement was posted as an inline finding and fixed in `91c75b86`; the thread is resolved. No remaining actionable finding in this review. This is author self-review, not CODEOWNER approval.

The combined regression command passed **469 tests in 109.85 s** at `0dd17f81`; the only subsequent change is the two-line documentation correction. Documentation validation passed on the new head (90 links/anchors and seven complete CLI configs). The repository Fast CI lint/format scope also passes locally.

```bash
python/aisimulate/.venv/bin/pytest -p no:timeout -q --disable-warnings \
  tests/test_cli_config.py tests/test_unified_cli.py tests/test_unified_traffic_runtime.py \
  tests/test_epd_cli.py tests/test_runner.py tests/e2e/test_unified_cli_engine.py \
  tests/sweeper/test_search_providers.py tests/sweeper/test_config.py \
  tests/sweeper/test_search_space.py tests/sweeper/test_sample.py \
  tests/sweeper/test_kv_load.py tests/sweeper/test_deploy.py \
  tests/sweeper/test_result.py tests/sweeper/test_search.py
```

The real H200-prefill/GB200-decode round trip completes six requests and reproduces candidate TTFT, TPOT and throughput. The tests also cover heterogeneous analytical EPD, inheritance and matching overrides, invalid domains/roles, backend-version disagreement, and mismatched Router hardware. Required hosted CI and CodeRabbit review remain separate gates.

**Review by simone-chen — APPROVED** [review]

File: docs/cli/user-guide.md [hunk 2]
```diff
@@ -919,8 +920,17 @@ engine:
...
 `optimization.hardware`. Every recommended prediction YAML replaces `auto` with that concrete
-identifier. Language-worker roles in aggregated or disaggregated mode use the same hardware;
-an optional analytical encoder pool can specify its own hardware.
+identifier. Language workers inherit that fallback unless a P/D role overrides it;
+an optional analytical encoder pool can also specify its own hardware.
+
+For heterogeneous P/D, set `engine.workers.prefill.hardware` and/or
+`engine.workers.decode.hardware`. An omitted role inherits `engine.hardware`. Both roles
+share the model, backend and backend version; when an override is present, an omitted
+version must resolve identically on both effective SKUs. Pin a common supported version
```
Comment (jasonqinzhou) [inline docs/cli/user-guide.md:929]: P2: The later language-worker paragraph (lines 972-974 on this head) still says all roles share hardware and that per-role overrides are rejected. That contradicts the new fields and this paragraph. Update the later statement to keep model/backend/version/context shared while allowing P/D hardware overrides; this also addresses the additional documentation finding in the CodeRabbit review.

File: python/aisimulate/src/aisimulate/config/engine.py [hunk 3]
```diff
@@ -390,6 +394,17 @@ def _validate_roles(self) -> EngineRecommendationConfig:
...
 
 
+def _validate_worker_hardware(*, modes: set[str], workers) -> None:
+    for role in ("aggregated", "prefill", "decode"):
+        worker = getattr(workers, role)
+        if worker is None or worker.hardware is None:
+            continue
+        if role == "aggregated" or "disaggregated" not in modes or "afd" in modes:
+            raise ValueError("worker hardware overrides require prefill/decode workers in disaggregated mode")
+        if not worker.hardware.strip() or worker.hardware == "auto":
```
[the remark this answers is not in the document]
Comment (jasonqinzhou) [reply python/aisimulate/src/aisimulate/config/engine.py]: Fixed in 0dba87b21f09159783ee3516c13d856df7f64e20. Worker hardware validation now rejects values with leading/trailing whitespace, including padded auto and concrete SKUs. Expanded the boundary matrix to both prefill/decode roles and prediction/recommendation schemas. All 230 focused config, EPD, native CLI round-trip, and Sweeper tests pass; Ruff and formatting pass.

File: docs/cli/user-guide.md (hunk 2 above)
Comment (jasonqinzhou) [reply docs/cli/user-guide.md:929]: Fixed in 0dba87b21f09159783ee3516c13d856df7f64e20. The later paragraph now distinguishes the shared model/backend/version/context fields from the P/D hardware fallback and overrides. The guide example also validates and exactly matches the real native round-trip fixture.

**Comment by jasonqinzhou** [comment]
Codex closeout self-review completed at `0dba87b21f09159783ee3516c13d856df7f64e20` against current `main` at `be8f69749589aea5a376083436e054d75810d379`.

Addressed the padded worker-hardware validation finding and the remaining contradictory language-worker documentation statement. Both conversations now identify the fixing commit and are resolved. Re-read the complete diff and traced configuration through Sweeper, saved prediction YAML, compiler timing/KV metadata, native replay, and the Router AIC hardware guard. No remaining actionable finding in this review.

Validation: **230 passed in 100.38 s** with an isolated repository-managed environment and a native extension built from this checkout:

```bash
python/aisimulate/.venv/bin/python -m pytest -q -p no:timeout \
  tests/test_cli_config.py tests/test_epd_cli.py \
  tests/e2e/test_unified_cli_engine.py tests/sweeper/test_search.py
```

Ruff, formatting, `git diff --check`, and guide-YAML validation passed. The heterogeneous guide YAML matches the native CLI fixture exactly. This is author self-review; Full CI and applicable approval remain separate merge gates.

File: python/aisimulate/src/aisimulate/config/engine.py [hunk 4]
```diff
@@ -390,6 +394,18 @@ def _validate_roles(self) -> EngineRecommendationConfig:
...
 
+def _validate_worker_hardware(*, modes: set[str], workers) -> None:
+    for role in ("aggregated", "prefill", "decode"):
+        worker = getattr(workers, role)
+        if worker is None or worker.hardware is None:
+            continue
+        if role == "aggregated" or "disaggregated" not in modes or "afd" in modes:
+            raise ValueError("worker hardware overrides require prefill/decode workers in disaggregated mode")
+        hardware = worker.hardware.strip()
+        if not hardware or hardware == "auto" or hardware != worker.hardware:
```
[the remark this answers is not in the document]
Comment (jasonqinzhou) [reply python/aisimulate/src/aisimulate/config/engine.py:405]: Fixed in 3c3a4fa342e3d50717cf95ca99e83a14fd5104cc. Confirmed the pinned-version path previously reached native KV-cache setup and produced a runtime error. The prediction compiler now rejects unknown explicit worker SKUs through load_system_spec before deployment compilation, including pinned backend versions. Both role/version boundary cases are covered, and tests preserve support for custom configured system paths. Final focused suite: 236 passed in 95.65 s; Ruff/format/whitespace checks pass. The real CLI now reports the invalid worker field as a configuration error with exit code 2.

**Comment by jasonqinzhou** [comment]
Codex re-review completed at `3c3a4fa342e3d50717cf95ca99e83a14fd5104cc` against current main `be8f69749589aea5a376083436e054d75810d379`.

Confirmed the new CodeRabbit finding with the public CLI: an unknown role SKU with a pinned backend version reached KV-cache setup and produced a runtime error. The compiler now checks explicit worker hardware through the configured system registry before building the deployment. This preserves custom registered SKUs and leaves omitted worker overrides unchanged.

The four new unknown-SKU boundary cases failed before the fix. Final focused regression: **236 passed in 95.65 s**, including both roles, pinned/implicit versions, custom system paths, native heterogeneous recommend/save/predict, and analytical EPD. The public CLI now exits 2 with `unknown workers.decode.hardware 'nonexistent_worker_sku': no system configuration found`. Ruff, formatting, and whitespace checks pass.

Reviewed the correction and its interactions with the complete feature diff; no remaining actionable finding. All addressed conversations are resolved. Updated-head CodeRabbit and Full CI remain required before the authorized merge.
